src/Features/CSharp/Portable/ConvertLinq/CSharpConvertForEachToLinqQueryProvider.cs

```diff
@@ -0,0 +1,585 @@
+                        statementsCannotBeConverted = Enumerable.Empty<StatementSyntax>();
+                        break;
+
+                    case SyntaxKind.EmptyStatement:
+                        statementsCannotBeConverted = Enumerable.Empty<StatementSyntax>();
+                        break;
+
+                    default:
+                        statementsCannotBeConverted = new[] { current };
+                        break;
+                }
+            }
+
+            return (convertingNodes.ToImmutableArray(), identifiers.ToImmutableArray(), statementsCannotBeConverted.ToImmutableArray());
+        }
+
+        protected override bool TryBuildSpecificConverter(
+            ForEachStatementSyntax forEachStatement,
+            SemanticModel semanticModel,
+            ImmutableArray<SyntaxNode> convertingNodes,
+            StatementSyntax statementCannotBeConverted,
+            CancellationToken cancellationToken,
+            out IConverter converter)
+        {
+            switch (statementCannotBeConverted.Kind())
+            {
+                case SyntaxKind.ExpressionStatement:
+                    var expression = ((ExpressionStatementSyntax)statementCannotBeConverted).Expression;
+                    switch (expression.Kind())
+                    {
+                        case SyntaxKind.PostIncrementExpression:
+                            // No matter what can be used as the last select statement for the case of Count. We use SyntaxFactory.IdentifierName(forEachStatement.Identifier).
+                            converter = new ConvertToCountSyntaxEditor(
+                                forEachStatement,
+                                convertingNodes,
+                                SyntaxFactory.IdentifierName(forEachStatement.Identifier),
+                                ((PostfixUnaryExpressionSyntax)expression).Operand);
+                            return true;
+
+                        case SyntaxKind.InvocationExpression:
+                            var invocationExpression = (InvocationExpressionSyntax)expression;
+                            if (invocationExpression.Expression is MemberAccessExpressionSyntax memberAccessExpression &&
+                                semanticModel.GetSymbolInfo(memberAccessExpression, cancellationToken).Symbol is IMethodSymbol methodSymbol &&
+                                IsList(methodSymbol.ContainingType, semanticModel) &&
+                                methodSymbol.Name.Equals(nameof(IList.Add)) &&
+                                methodSymbol.Parameters.Length == 1)
+                            {
+                                converter = new ConvertToListSyntaxEditor(
+                                    forEachStatement,
+                                    convertingNodes,
+                                    invocationExpression.ArgumentList.Arguments.Single().Expression,
```

Comment:
you're not guaranteed to have a single argument (thoug it would be weird).  to be safe, add a length check in the if-condition.

Reply:
I believe I check that invocationExpression is in the form of 'something.Add(someItem)'
including that the method is 'Add' and it has a single parameter.
Is not this enough?


---
In reply to: [191058316](https://github.com/dotnet/roslyn/pull/27130#discussion_r191058316) [](ancestors = 191058316)

Reply:
> I believe I check that invocationExpression is in the form of 'something.Add(someItem)'

Where do you check that?

Reply:
methodSymbol.Name.Equals(nameof(IList.Add)) &&
                                methodSymbol.Parameters.Length == 1)

---
In reply to: [194904413](https://github.com/dotnet/roslyn/pull/27130#discussion_r194904413) [](ancestors = 194904413)

Reply:
That is only validating the method symbol.  It does not guarantee that hte *syntax* also matches.  Error conditions/etc. may make this not true.

Reply:
OK. Thanks! We check that there are no errors, but let us be more defensive. Adding 'invocationExpression.ArgumentList.Arguments.Count == 1'

---
In reply to: [194911675](https://github.com/dotnet/roslyn/pull/27130#discussion_r194911675) [](ancestors = 194911675)